keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+4. Proceed with the maintenance and shut down or restart the node.
+5. On platforms and nodes that support it, the kubelet will try to detect the imminent shutdown and
+   then attempt to perform a [Graceful Node Shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown):
+   - delay the shutdown pending graceful termination of remaining pods
+   - terminate remaining pods in reverse priority order (see [pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))
+
+The main problem is that the current approach tries to solve this in an application agnostic way
+and will simply attempt to remove all the pods currently running on the node. Since this approach
+cannot be applied generically to all pods, the Kubernetes project has defined special
+[drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162)
+that either skip groups of pods or an admin has to consent to override those groups to be either
+skipped or deleted. This means that without knowledge of all the underlying applications on the
+cluster, the admin has to make a potentially harmful decision.
+
+From an application owner or developer perspective, the only standard tool they have is
+a PodDisruptionBudget. This is sufficient in a basic scenario with a simple multi-replica
+application. The edge case applications, where this does not work are very important to
+the cluster admin, as they can block the node drain. And, in turn, very important to the
+application owner, as the admin can then override the pod disruption budget and disrupt their
+sensitive application anyway.
+
+List of cases where the current solution is not optimal:
+
+1. Without extra manual effort, an application running with a single replica has to settle for
+   experiencing application downtime during the node drain. They cannot use PDBs with
+   `minAvailable: 1` or `maxUnavailable: 0`, or they will block node maintenance. Not every user
+   needs high availability either, due to a preference for a simpler deployment model, lack of
+   application support for HA, or to minimize compute costs. Also, any automated solution needs
+   to edit the PDB to account for the additional pod that needs to be spun to move the workload
+   from one node to another. This has been discussed in the issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
+   and in the issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
+2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable
+   number of pods; for example applications with a configured horizontal pod autoscaler (HPA). These
```

Comment:
The two possible alternative would be:
1. extend HPA to update associated PDB to maintain the required minimum? 
2. every application should have a reasonable minimum which would allow the application to survive any kind of disruption (planned or unplanned).

Reply:
> extend HPA to update associated PDB to maintain the required minimum?

- This is not possible when n=1, which is the core of the https://github.com/kubernetes/kubernetes/issues/93476 issue.
- Although it is possible for n>1, I would be afraid of conflicts when introducing such feature. PDB was not introduced as a controlled/changing object like for example ReplicaSets. It is set by operators/controllers, manually or as part of various devops flows. 
- There is no backlink from pod/HPA to a PDB. Even if we introduce such a backlink, there are still problems. What do we do when the PDB does not exist and there is one specified by a HPA. Should it be the responsibility of the HPA to become an operator of a PDB (or become other type of a PDB)?  We could run again into the conflicts with existing flows as seen in the previous point.
- There is no mapping from a PDB to an application, there is only a selector. Although not recommended, there could be two applications under one PDB. One controlled by HPA and the other one not. Or any combination of these.

> every application should have a reasonable minimum which would allow the application to survive any kind of disruption (planned or unplanned).

This is not always possible/desired, as seen in https://github.com/kubernetes/kubernetes/issues/66811, https://github.com/kubernetes/kubernetes/issues/114877 and https://github.com/kubernetes/kubernetes/issues/93476